File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
// TC- O(n)
// SC- O(n)

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int maxArea = 0;
        stack<int> st; // stores indices
        heights.push_back(0); // to flush out stack at the end

        for(int i=0; i<heights.size(); i++){
            while(not st.empty() and heights[st.top()]>heights[i]){
                int height = heights[st.top()];
                st.pop();
                int width;
                int right_boundary = i-1;
                int left_boundary = st.empty() ? 0:st.top()+1;
                width = right_boundary-left_boundary+1;
                maxArea = max(maxArea, height*width);
            }
            st.push(i);
        }
        return maxArea;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
# TC- O(n)
# SC- O(n)

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        max_area = 0
        stack = [] # store indices 
        heights.append(0) # to flush out stack

        for idx, ht in enumerate(heights):
            while stack and ht<heights[stack[-1]]:
                popped_idx = stack.pop()
                height = heights[popped_idx]
                left_smaller = stack[-1] if stack else -1
                width = idx - left_smaller -1 
                area = height*width
                max_area = max(area, max_area)
            stack.append(idx)

        return max_area